src/Bedrock.Framework/Protocols/MessagePipeReader.cs

```diff
@@ -0,0 +1,155 @@
+            {
+                _backlog.Consume(consumedLength);
+            }
+            else
+            {
+                var unconsumed = _message.Slice(consumed);
+                if (!unconsumed.IsEmpty)
+                {
+                    foreach (var m in unconsumed)
+                    {
+                        _backlog.Write(m.Span);
+                    }
+                }
+            }
+
+            // REVIEW: Use the correct value for examined
+            _reader.AdvanceTo(_consumed, _examined);
+        }
+
+        public override void CancelPendingRead()
+        {
+            _reader.CancelPendingRead();
+        }
+
+        public override void Complete(Exception exception = null)
+        {
+            _reader.Complete();
+        }
+
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            while (true)
+            {
+                var result = await _reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+
+                if (TryCreateReadResult(result, out var readResult))
+                    return readResult;
+            }
+        }
+
+        public override bool TryRead(out ReadResult readResult)
+        {
+            if (_reader.TryRead(out var result))
+            {
+                if (TryCreateReadResult(result, out readResult))
+                    return true;
+            }
+
+            if (_allExamined)
+            {
+                readResult = default;
+                return false;
+            }
+
+            _message = new ReadOnlySequence<byte>(_backlog.WrittenMemory);
+            readResult = new ReadResult(_message, _isCanceled, _isCompleted);
+            return true;
+        }
+
+        private bool TryCreateReadResult(ReadResult underlyingReadResult, out ReadResult readResult)
+        {
+            var buffer = underlyingReadResult.Buffer;
+            _isCanceled = underlyingReadResult.IsCanceled;
+            _isCompleted = underlyingReadResult.IsCompleted;
+
+            if (_isCanceled)
+            {
+                readResult = new ReadResult(default, _isCanceled, _isCompleted);
+                return true;
+            }
+
+            if (_messageReader.TryParseMessage(buffer, ref _consumed, ref _examined, out _message))
+            {
+                if (_backlog.UnconsumedWrittenCount > 0)
+                {
+                    foreach (var m in _message)
+                    {
+                        _backlog.Write(m.Span);
+                    }
+
+                    _message = new ReadOnlySequence<byte>(_backlog.WrittenMemory);
+                }
+
+                if (!_message.IsEmpty)
+                    _allExamined = false;
+
+                readResult = new ReadResult(_message, _isCanceled, _isCompleted);
+                return true;
+            }
+
+            if (_isCompleted)
+            {
+                if (!buffer.IsEmpty)
```

Comment:
Maybe we don't throw here and let the caller handle this situation.

Reply:
How would we signal this occured?
Namely how would the caller differentiate between:

1. It completed and successfully parsed the message
2. It completed without returning any bytes
3. It completed and could not parse the message (this case).